File: backend/python/ethnos/cartas_visiveis.py
from __future__ import annotations
from .carta import Carta


class CartasVisiveis:
    """
    Área compartilhada de cartas que todos os jogadores podem recrutar (RF10).

    Recebe cartas descartadas da mão (RF16) e dragões revelados do deck (RF11).
    É esvaziada ao fim de cada era (RF20).
    """

    def __init__(self) -> None:
        self._cartas: list[Carta] = []

    def adicionar(self, carta: Carta) -> None:
        self._cartas.append(carta)

    def adicionar_varias(self, cartas: list[Carta]) -> None:
        self._cartas.extend(cartas)

    def retirar(self, idx: int) -> Carta:
        if idx < 0 or idx >= len(self._cartas):
            raise ValueError("Índice inválido para cartas visíveis.")
        return self._cartas.pop(idx)

    def limpar(self) -> None:
        self._cartas.clear()

    def __len__(self) -> int:
        return len(self._cartas)

    def __iter__(self):
        return iter(self._cartas)

    def __repr__(self) -> str:
        return f"CartasVisiveis({len(self._cartas)} cartas)"

File: backend/ts/cartasVisiveis.ts
import { Carta } from './carta';

/**
 * Área compartilhada de cartas que todos os jogadores podem recrutar (RF10).
 *
 * Recebe cartas descartadas da mão (RF16) e dragões revelados do deck (RF11).
 * É esvaziada ao fim de cada era (RF20).
 */
export class CartasVisiveis {
  private _cartas: Carta[];

  constructor() {
    this._cartas = [];
  }

  adicionar(carta: Carta): void {
    this._cartas.push(carta);
  }

  adicionarVarias(cartas: Carta[]): void {
    this._cartas.push(...cartas);
  }

  retirar(idx: number): Carta {
    if (idx < 0 || idx >= this._cartas.length) {
      throw new Error('Índice inválido para cartas visíveis.');
    }
    return this._cartas.splice(idx, 1)[0];
  }

  limpar(): void {
    this._cartas = [];
  }

  get length(): number {
    return this._cartas.length;
  }

  [Symbol.iterator](): Iterator<Carta> {
    return this._cartas[Symbol.iterator]();
  }

  toArray(): Carta[] {
    return [...this._cartas];
  }

  toString(): string {
    return `CartasVisiveis(${this._cartas.length} cartas)`;
  }
}
